cabal-install/src/Distribution/Client/CmdConfigure.hs

```diff
@@ -140,9 +149,101 @@ configureAction flags@NixStyleFlags {..} _extraArgs globalFlags = do
+  commandSynopsis     = "Add extra project configuration",
+  commandUsage        = usageAlternatives "v2-reconfigure" [ "[FLAGS]" ],
+  commandDescription  = Just $ \_ -> wrapText $
+        "Adjust how the project is built by setting additional package flags "
+     ++ "and other flags.\n\n"
+
+     ++ "The reconfiguration options are written to the 'cabal.project.local' "
+     ++ "file (or '$project_file.local', if '--project-file' is specified) "
+     ++ "which extends the configuration from the 'cabal.project' file "
+     ++ "(if any).\n\n"
+
+     ++ "The v2-reconfigure command also checks that the project configuration "
+     ++ "will work. In particular it checks that there is a consistent set of "
+     ++ "dependencies for the project as a whole.\n\n"
+
+     ++ "It is never necessary to use the 'v2-reconfigure' command. It is "
+     ++ "merely a convenience in cases where you do not want to specify flags "
+     ++ "to 'v2-build' (and other commands) every time and yet do not want "
+     ++ "to alter the 'cabal.project' persistently.",
+  commandNotes        = Just $ \pname ->
+        "Examples:\n"
+     ++ "  " ++ pname ++ " v2-reconfigure --with-compiler ghc-7.10.3\n"
+     ++ "    Adjust the project configuration to use the given compiler\n"
+     ++ "    program and check the resulting configuration works.\n"
+     ++ "  " ++ pname ++ " v2-reconfigure\n"
+     ++ "    Reset the local configuration to empty and check the overall\n"
+     ++ "    project configuration works.\n"
+
+  , commandDefaultFlags = defaultNixStyleFlags ()
+  , commandOptions      = filter (\o -> optionName o /= "ignore-project")
+                        . nixStyleOptions (const [])
+  }
+
+reconfigureAction :: NixStyleFlags () -> [String] -> GlobalFlags -> IO ()
+reconfigureAction flags extraArgs globalFlags = do
+  reconf <- reconfigureAction' flags extraArgs globalFlags
+  let go (baseCtx, projConfig) = 
+        writeProjectLocalExtraConfig (distDirLayout baseCtx) projConfig
+
+  either go go reconf
+
+reconfigureAction' 
+    :: NixStyleFlags () 
+    -> [String] 
+    -> GlobalFlags 
+    -> IO (Either (ProjectBaseContext, ProjectConfig) (ProjectBaseContext, ProjectConfig))
+reconfigureAction' flags@NixStyleFlags {..} extraArgs globalFlags = do
+  baseCtx <- establishProjectBaseContext v cliConfig OtherCommand
+
+  let localFile = distProjectFile (distDirLayout baseCtx) "local"
+
+  exists <- doesFileExist localFile
+  if not exists
+    then do
+      notice v $
+        quote (takeFileName localFile)
+        <> " doesn't exist, calling configure instead."
+
+      -- @configureAction'@ is called, as opposed to @configureAction@, to avoid
+      -- calling on the writing function for configure, thusly maintaining the 
+      -- separation on the behaviour that modify files
+      Left <$> configureAction' flags extraArgs globalFlags
+  
+    else do
+      conf <- runRebuild (distProjectRootDirectory . distDirLayout $ baseCtx) $
```

Comment:
What does running a rebuild do here?

Reply:
Nothing really, it's just so `readProjectLocalExtraConfig` is wrapped by the `Rebuild` monad.